lamagraph-core/src/Core/MemoryManager.hs

```diff
@@ -0,0 +1,383 @@
+  { _secP :: Vec portsNumber (Maybe (Port portsNumber))
+  , _primeP :: Maybe (Port portsNumber)
+  }
+  deriving (Show, Eq, Generic, NFDataX, Default)
+
+$(makeLenses ''NodePortsInfo)
+
+-- | Update `Port`s of `Node` by `NodePortsInfo`
+updateNode ::
+  (KnownNat portsNumber) =>
+  Node portsNumber ->
+  NodePortsInfo portsNumber ->
+  Node portsNumber
+updateNode oldNode (NodePortInfo maybeSecPortsAddr maybePrimaryPort) =
+  set primaryPort newPrimPort (set secondaryPorts newSecPorts oldNode)
+ where
+  newSecPorts =
+    zipWith
+      (<|>)
+      maybeSecPortsAddr
+      (oldNode ^. secondaryPorts)
+  newPrimPort = case maybePrimaryPort of
+    Nothing -> oldNode ^. primaryPort
+    Just primPort -> primPort
+
+-- | The flag to distinguish between nodes that need to be loaded from ram and local ones
+data LocalFlag
+  = Local
+  | External
+  deriving (Eq)
+
+{- | Update ports info in attached `Node`.
+It accumulates in key-value map, where key is `ActualAddressNum` and value is a pair of `NodePortsInfo` and `LocalFlag` to distinguish local and external nodes.
+The first one we can upload from locals and another one we have to upload from ram
+-}
+updatePortsInfoByPort ::
+  forall portsNumber maxNumOfChangedNodes.
+  (KnownNat portsNumber, KnownNat maxNumOfChangedNodes) =>
+  Map maxNumOfChangedNodes (NodePortsInfo portsNumber, LocalFlag) ->
+  LoadedNode portsNumber ->
+  Port portsNumber ->
+  IdOfPort portsNumber ->
+  Map maxNumOfChangedNodes (NodePortsInfo portsNumber, LocalFlag)
+updatePortsInfoByPort infoVec localNodeWithAddress (Port maybeAddr connectedToPortId) portId =
+  case maybeAddr of
+    Nothing -> infoVec -- or @error "Port must be connected"@
+    Just addr -> case addr of
+      LocalAddress localAddressNum -> update Local localAddressNum
+      ActualAddress addressNum -> update External addressNum
+     where
+      update localFlag addressNum =
+        insertWith
+          infoVec
+          ( \case
+              Nothing -> Just (constructNewInfo def, localFlag)
+              Just (info, _) -> Just (constructNewInfo info, localFlag)
+          )
+          addressNum
+       where
+        newPort = Port (Just $ ActualAddress (localNodeWithAddress ^. originalAddress)) portId
+        constructNewInfo info@(NodePortInfo secPortsInfo _) = case connectedToPortId of
+          Primary -> set primeP (Just newPort) info
+          Id i -> set secP (replace i (Just newPort) secPortsInfo) info
+
+updatePortsInfoByEdge ::
+  forall portsNumber maxNumOfChangedNodes.
+  (KnownNat portsNumber, KnownNat maxNumOfChangedNodes) =>
+  Map maxNumOfChangedNodes (NodePortsInfo portsNumber, LocalFlag) ->
+  Edge portsNumber ->
+  Map maxNumOfChangedNodes (NodePortsInfo portsNumber, LocalFlag)
+updatePortsInfoByEdge oldInfoVec (Edge leftE rightE) =
+  update
+    rightAddrNum
+    leftAddrNum
+    (rightE ^. idOfPort)
+    (leftE ^. idOfPort)
+    $ update leftAddrNum rightAddrNum (leftE ^. idOfPort) (rightE ^. idOfPort) oldInfoVec
```

Comment:
Can you please elaborate on this function? Especially chained call to `update` with mirrored arguments weirded me out.